kolibri/core/bookmarks/api.py

```diff
@@ -0,0 +1,74 @@
+from rest_framework.permissions import IsAuthenticated
+from rest_framework.response import Response
+from rest_framework.status import HTTP_200_OK
+from rest_framework.status import HTTP_201_CREATED
+from rest_framework.status import HTTP_204_NO_CONTENT
+from rest_framework.status import HTTP_400_BAD_REQUEST
+from rest_framework.status import HTTP_403_FORBIDDEN
+from rest_framework.status import HTTP_405_METHOD_NOT_ALLOWED
+
+from .models import Bookmark
+from .serializers import BookmarksSerializer
+from kolibri.core.api import ValuesViewset
+
+
+class BookmarksViewSet(ValuesViewset):
+    values = ("channel_id", "contentnode_id", "id", "content_id")
+    serializer_class = BookmarksSerializer
+    permission_classes = [IsAuthenticated]
+
+    def list(self, request):
+        bookmarks = Bookmark.objects.filter(facility_user=request.user).values(
+            "channel_id", "content_id", "contentnode_id", "id"
+        )
+        serializer = BookmarksSerializer(bookmarks, many=True)
+
+        return Response(serializer.data, status=HTTP_200_OK)
+
+    def create(self, request):
+        """
+        Overridden for two reasons:
+        1) Inject user into the data before creating it to avoid NOT NULL on facility_user
+        2) Provide meaningful HTTP responses accordingly, including 200 for duplicates
+        """
+        # POST requires these three to come from the client
+        required_keys = ["channel_id", "contentnode_id", "content_id"]
```

Comment:
Is it possible to just require the `contentnode_id`, since the others can be inferred from that ID? Also, this might be the only opportunity to do some validation that the content exists, since we don't have a ForeignKey relationship between Bookmarks and ContentNodes

Reply:
This seems reasonable, can simplify the serializer, do validation just that this specific node id exists, and then record all three values for future reference.

Reply:
No but I neither tested for that nor implemented it -_-